pkg/restore/ingester.go

```diff
@@ -0,0 +1,573 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package restore
+
+import (
+	"bytes"
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/kvproto/pkg/errorpb"
+	sst "github.com/pingcap/kvproto/pkg/import_sstpb"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/codec"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/keepalive"
+
+	"github.com/pingcap/br/pkg/conn"
+	berrors "github.com/pingcap/br/pkg/errors"
+	"github.com/pingcap/br/pkg/kv"
+	"github.com/pingcap/br/pkg/utils"
+)
+
+const (
+	dialTimeout = 5 * time.Second
+
+	gRPCKeepAliveTime    = 10 * time.Second
+	gRPCKeepAliveTimeout = 3 * time.Second
+
+	// See: https://github.com/tikv/tikv/blob/e030a0aae9622f3774df89c62f21b2171a72a69e/etc/config-template.toml#L360
+	regionMaxKeyCount = 1440000
+)
+
+type retryType int
+
+const (
+	retryNone retryType = iota
+	retryWrite
+	retryIngest
+)
+
+type gRPCConns struct {
+	mu    sync.Mutex
+	conns map[uint64]*conn.Pool
+}
+
+func (conns *gRPCConns) Close() {
+	conns.mu.Lock()
+	defer conns.mu.Unlock()
+
+	for _, cp := range conns.conns {
+		cp.Close()
+	}
+}
+
+// Ingester writes and ingests kv to TiKV.
+// which used for both BR log restore and Lightning local backend.
+type Ingester struct {
+	// commit ts appends to key in tikv
+	TS uint64
+
+	conns gRPCConns
+
+	splitCli   SplitClient
+	workerPool *utils.WorkerPool
+
+	batchWriteKVPairs int
+	regionSplitSize   int64
+	tcpConcurrency    int
```

Comment:
I think it's better to put this field into `gRPCCoons` since it means the capacity of the connection pool

Reply:
Done